backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchKnnPredicate.java

```diff
@@ -45,6 +46,12 @@ private ElasticsearchKnnPredicate(AbstractKnnBuilder<?> builder) {
 		builder.vector = null;
 	}
 
+	protected JsonObject prepareFilter(PredicateRequestContext context) {
+		JsonObject mainFilter = filter == null ? null : filter.toJsonQuery( context );
+		JsonArray filters = context.tenantAndRoutingFilters();
+		return Queries.boolCombineMust( mainFilter, filters );
```

Comment:
This works OK for the regular case when knn is not nested. 
As soon as we add a knn predicate to a nested one ... 😞 the results are all over the place between the backends..
for Lucene the results seem fine, with Elasticsearch -- it sometimes works, sometimes it fails with 
```json
{
        "type": "illegal_state_exception",
        "reason": "Parent query must not match any docs besides parent filter. Combine them as must (+) and must-not (-) clauses to find a problem doc. docID\u003d8"
      }
```
and OpenSearch "ignores" the tenant and routes i.e. it doesn't fail, but it doesn't apply the filtering correctly. 

Soooo... I'm thinking if it would make sense to include the tenant id and route key in the nested objects so we can use them for filtering.  would it make sense to go this route, or maybe you have some other idea to test?

Reply:
I thought about logging/throwing an exception if we need to apply tenant/route filters and the knn is nested... but ... didn't want to add new limitations for this predicate as we've just removed some of them 😞

Reply:
> I'm thinking if it would make sense to include the tenant id and route key in the nested objects so we can use them for filtering

You mean automatically adding the tenant ID/routing filter to `nested` predicates?

I wouldn't have expected that to be necessary, but... if it is, why not. I don't think it will hurt.

Reply:
not really 🙈 😃. Initially, I thought that this should work for a nested query:

```json
{
  "query": {
    "bool": {
      "must": {
        "nested": {
          "path": "nestedObject",
          "query": {
            "knn": {
              "field": "nestedObject.floats",
              "num_candidates": 1,
              "query_vector": [
                1000.0,
                1000.0
              ]
            }
          }
        }
      },
      "filter": [
        {
          "term": {
            "_tenant_id": "tenant_1"
          }
        }
      ]
    }
  }
}
```
Having a filter at the root level and no additional filter in the knn that lives in the nested query... but that does not work, meaning having two docs:
```
1. tenant: tenant_1   nestedObject.floats: [10.0, 10.0]
2. tenant: tenant_2   nestedObject.floats: [1000.0, 1000.0]
```
and running that search query for `tenant_1` results in ... no matches 😞 
what I was considering: adding a "system" field to `nestedObject` -- `nestedObject._tenant_id` so that we'd be able to say:
```json
{
  "query": {
    "bool": {
      "must": {
        "nested": {
          "path": "nestedObject",
          "query": {
            "knn": {
              "field": "nestedObject.floats",
              "num_candidates": 1,
              "query_vector": [
                1000.0,
                1000.0
              ],
              "filter": [
                {
                  "term": {
                    "nestedObject._tenant_id": "tenant_1"
                  }
                }
              ]
            }
          }
        }
      },
      "filter": [
        {
          "term": {
            "_tenant_id": "tenant_1"
          }
        }
      ]
    }
  }
}
```

Reply:
That's... reasonable. And also an annoying breaking change :D 

I think in the short term your only option is to document that with Elasticsearch/OpenSearch, knn inside nested fields with multitenancy is not supported.

You might try to report this as a bug to Elasticsearch/OpenSearch; too. It seems strange that in the context of a nested query, the knn query doesn't apply only to children matched by the parent query. But I think this is unlikely to get wide attention, unfortunately...